test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -198,30 +239,37 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
 		machinePool.Spec.Replicas = pointer.Int32(1)
 	}
 
-	pool, err := docker.NewNodePool(ctx, r.Client, cluster, machinePool, dockerMachinePool)
-	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to build new node pool")
+	// First, reconcile the Docker containers, but do not delete any as we need to delete the Machine to ensure node cordon/drain.
+	// Similarly, providers implementing MachinePool Machines will need to reconcile their analogous infrastructure instances (aside
+	// from deletion) before reconciling InfraMachinePoolMachines.
+	if err := r.reconcileDockerContainers(ctx, cluster, machinePool, dockerMachinePool); err != nil {
+		return ctrl.Result{}, err
 	}
 
-	// Reconcile machines and updates Status.Instances
-	remoteClient, err := r.Tracker.GetClient(ctx, client.ObjectKeyFromObject(cluster))
-	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to generate workload cluster client")
+	// Second, reconcile the DockerMachines once the Docker containers are created. This means creating a DockerMachine for each newly
+	// created Docker container as well as handling container deletion. Instead of deleting an infrastructure instance directly, we want to reach up
+	// to pull the associated DockerMachine and its owner Machine and call the deletion the owner Machine, which will trigger cordon and drain as well
+	// as delete the DockerMachine, which then deletes the Docker container.
+	// Similarly, providers will need to create InfraMachines for each instance, and instead of deleting instances directly, delete the associated Machine.
+	if err := r.reconcileDockerMachines(ctx, cluster, machinePool, dockerMachinePool); err != nil {
+		return ctrl.Result{}, err
 	}
-	res, err := pool.ReconcileMachines(ctx, remoteClient)
+
+	// Refresh the list of DockerMachines to ensure the provider IDs are up to date.
```

Comment:
is this really a refresh or are we fetching the list for the first time?

Reply:
Technically it is a refresh since we got the list of DockerMachines in `reconcileDockerMachines()`, and I wanted to answer the question of why we're not just returning that list from the function. But I can update the comment to make it a bit more clear.